### PR TITLE
Migrate picard to json API

### PR DESCRIPTION
**Pull request by samj1912**
PR up for review.
Will add tests soon.

### CODE REVIEW

File: picard/webservice/api_helpers.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def _find(self, entitytype, handler, **kwargs):
                 value = escape_lucene_query(value).strip().lower()
                 if value:
                     query.append('%s:(%s)' % (name, value))
-            query = ' '.join(query)
+            query = ''.join(query)
```
Comment (samj1912) [inline picard/webservice/api_helpers.py]: I saw the urls, I dont think the spaces are needed between query args?

Comment (zas) [inline picard/webservice/api_helpers.py]: Well, since this change is unrelated, just revert it for now.

File: picard/album.py [hunk 2]
```diff
@@ -182,7 +182,7 @@ def _release_request_finished(self, document, http, error):
             else:
                 try:
                     parsed = self._parse_release(document)
-                except:
+                except Exception as e:
```
Comment (zas) [inline picard/album.py]: `e` is unused here

File: picard/coverart/providers/caa.py [hunk 3]
```diff
@@ -242,10 +241,10 @@ def _has_suitable_artwork(self):
 
         if not caa_has_suitable_artwork:
```
Comment (mineo) [inline picard/coverart/providers/caa.py]: `caa_hash_suitable_artwork` can now be not defined at this point.

File: picard/metadata.py [hunk 4]
```diff
@@ -174,41 +174,40 @@ def compare_to_release_parts(self, release, weights):
...
 
         return parts
 
     def compare_to_track(self, track, weights):
         parts = []
-
         if 'title' in self:
             a = self['title']
-            b = track.title[0].text
+            b = track['title']
```
Comment (mineo) [inline picard/metadata.py]: I just got a TypeError here:

```
Traceback (most recent call last):
  File "./picard/webservice/__init__.py", line 266, in _process_reply
    self._handle_reply(reply, request, handler, parse_response_type, refresh)
  File "./picard/webservice/__init__.py", line 255, in _handle_reply
    handler(document, reply, error)
  File "./picard/acoustid/__init__.py", line 93, in _on_lookup_finished
    next_func(doc, http, error)
  File "./picard/file.py", line 599, in _lookup_finished
    reverse=True, key=itemgetter(0))[0]
  File "./picard/file.py", line 598, in <genexpr>
    track, self.comparison_weights) for track in tracks),
  File "./picard/metadata.py", line 194, in compare_to_track
    b = track['title']
TypeError: 'NoneType' object is not subscriptable
```

I was trying to load https://musicbrainz.org/release/772ae1d4-4fa8-4031-83b0-37c8f38df166.

File: picard/acoustid/__init__.py [hunk 5]
```diff
@@ -23,7 +23,10 @@
 from picard import config, log
 from picard.const import FPCALC_NAMES
 from picard.util import find_executable
-from picard.util.xml import XmlNode
+from picard.acoustid.json_helpers import parse_recording
+
+# Not an unused import. Don't remove it
```
Comment (mineo) [inline picard/acoustid/__init__.py]: What is this import used for? Only the import line in `tagger.py`?

Comment (samj1912) [inline picard/acoustid/__init__.py]: Yes

**Comment by zas** [comment]
> Also, is the JSON web service now considered stable? https://beta.musicbrainz.org/doc/Development/JSON_Web_Service still has a big fat warning and https://tickets.metabrainz.org/browse/MBS-3072 mentions at least one open issue in the comments.

It should, but it wasn't tested enough imho, this move is done with the help of ws devs, we'll fix issues related to json ws in priority.
Also we want to create a so-called ws 2.5, which is basically ws 2 but json only and with api key, so we need the json ws 2 to work as well as xml one. So all this fits our plans.

**Comment by mineo** [comment]
Another thing: on master, all release types (in `~primaryreleasetype`, `~secondaryreleasetype` and `releasetype` are lower-cased, with this PR it's only the primary one,

File: picard/coverart/providers/caa.py (hunk 3 above)
Comment (samj1912) [inline picard/coverart/providers/caa.py]: fixed

File: picard/metadata.py [hunk 6]
```diff
@@ -135,10 +134,11 @@ def compare_to_release_parts(self, release, weights):
             except ValueError:
                 pass
             else:
-                if "title" in weights:
-                    b = int(release.medium_list[0].medium[0].track_list[0].count)
-                else:
-                    b = int(release.medium_list[0].track_count[0].text)
+                if 'media' in release:
+                    if "title" in weights:
+                        b = int(release['media'][0]['track-count'])
```
Comment (samj1912) [inline picard/metadata.py]: fixed

File: picard/coverart/providers/caa.py [hunk 7]
```diff
@@ -202,24 +202,23 @@ def __init__(self, coverart):
...
-                      % self.release.id)
+                      % self.release['id'])
             return False
 
         if self.restrict_types:
             want_front = 'front' in self.caa_types
             want_back = 'back' in self.caa_types
-            caa_has_front = caa_node.front[0].text == 'true'
-            caa_has_back = caa_node.back[0].text == 'true'
+            caa_has_front = caa_node['front']
```
Comment (samj1912) [inline picard/coverart/providers/caa.py]: yes

File: picard/collection.py [hunk 8]
```diff
@@ -106,23 +106,24 @@ def request_finished(document, reply, error):
                 echo=log.error
             )
             return
-        collection_list = document.metadata[0].collection_list[0]
-        if "collection" in collection_list.children:
+        if document and "collection" in document:
+            collection_list = document['collections']
```
Comment (samj1912) [inline picard/collection.py]: not sure what you mean?

**Comment by samj1912** [comment]
> Another thing: on master, all release types (in ~primaryreleasetype, ~secondaryreleasetype and releasetype are lower-cased, with this PR it's only the primary one,

Fixed

File: picard/collection.py (hunk 8 above)
Comment (zas) [inline picard/collection.py]: Ignore ;)

File: picard/album.py [hunk 9]
```diff
@@ -232,21 +231,21 @@ def _finalize_loading(self, error):
...
                     discpregap = True
                     absolutetracknumber += 1
-                    track = self._finalize_loading_track(medium_node.pregap[0], mm, artists, va, absolutetracknumber, discpregap)
+                    track = self._finalize_loading_track(medium_node['pregap'], mm, artists, va, absolutetracknumber, discpregap)
                     track.metadata['~pregap'] = "1"
 
-                tracklist_node = medium_node.track_list[0]
-                track_count = int(tracklist_node.count)
+                tracklist_node = medium_node['tracks']
+                track_count = int(medium_node['track-count'])
```
Comment (mineo) [inline picard/album.py]: Are all of the `int` casts still required?

File: picard/collection.py [hunk 10]
```diff
@@ -106,23 +106,24 @@ def request_finished(document, reply, error):
                 echo=log.error
             )
             return
-        collection_list = document.metadata[0].collection_list[0]
-        if "collection" in collection_list.children:
+        if document and "collection" in document:
```
Comment (mineo) [inline picard/collection.py]: Is `document` ever a falsy value if `error` is not a truthy one? That would be really weird.

File: picard/mbjson.py [hunk 11]
```diff
@@ -363,7 +364,7 @@ def medium_to_metadata(node, m):
 
 
 def artist_to_metadata(node, m):
-    """Make meatadata dict from a XML 'artist' node."""
+    """Make meatadata dict from a json 'artist' node."""
```
Comment (mineo) [inline picard/mbjson.py]: Nitpicking: json is usually written as JSON.

File: picard/mbjson.py [hunk 12]
```diff
@@ -439,20 +440,22 @@ def release_group_to_metadata(node, m, release_group=None):
...
+            key = tag['name']
+            count = int(tag['count'])
+            obj.add_folksonomy_tag(key, count)
 
 
 def add_user_folksonomy_tags(node, obj):
-    for tag in node:
-        key = tag['name']
-        obj.add_folksonomy_tag(key, 1)
+    if obj:
```
Comment (mineo) [inline picard/mbjson.py]: Since we're following PEP8: these comparisons should be `if obj is None`.

File: picard/metadata.py [hunk 13]
```diff
@@ -135,10 +134,11 @@ def compare_to_release_parts(self, release, weights):
             except ValueError:
                 pass
             else:
-                if "title" in weights:
-                    b = int(release.medium_list[0].medium[0].track_list[0].count)
-                else:
-                    b = int(release.medium_list[0].track_count[0].text)
+                if 'media' in release:
```
Comment (mineo) [inline picard/metadata.py]: Is `media` ever not present here? Given that `totaltracks` is known at this point, this does look weird.

File: picard/metadata.py [hunk 14]
```diff
@@ -174,54 +174,56 @@ def compare_to_release_parts(self, release, weights):
...
+        if release['id'] in rg.loaded_albums:
             parts.append((1.0, 6))
 
         return parts
 
     def compare_to_track(self, track, weights):
         parts = []
+        result = (-1,)
+
+        if not track:
```
Comment (mineo) [inline picard/metadata.py]: When's that the case?

File: picard/ui/searchdialog.py [hunk 15]
```diff
@@ -645,19 +645,20 @@ def _cover_downloaded(self, row, data, http, error):
             try:
                 pixmap.loadFromData(data)
                 cover_cell.update(pixmap)
-            except:
+            except Exception as e:
                 cover_cell.not_found()
+                log.debug(e)
```
Comment (mineo) [inline picard/ui/searchdialog.py]: Why not `log.error`?

File: picard/acoustid/__init__.py (hunk 5 above)
Comment (mineo) [inline picard/acoustid/__init__.py]: You could always change the line in `tagger.py` :-)

Comment (samj1912) [inline picard/acoustid/__init__.py]: I did change the line in tagger. Just thought it was better to keep all the acoustid imports in one module.

File: picard/metadata.py (hunk 14 above)
Comment (samj1912) [inline picard/metadata.py]: When we are using the search dialog

File: picard/collection.py (hunk 10 above)
Comment (samj1912) [inline picard/collection.py]: I remember getting an error during one of my testing runs on this. That's why I made the change. Probably in case of an error or api limit reached or when the reply was not parsed correctly.

File: picard/metadata.py (hunk 14 above)
Comment (samj1912) [inline picard/metadata.py]: Sometimes the calling iterator has no tracks, thus causing an error here.

File: picard/album.py (hunk 9 above)
Comment (samj1912) [inline picard/album.py]: Just keeping them in case we switch back to strings for these. (I think there were some tickets related to this)

File: picard/metadata.py (hunk 13 above)
Comment (samj1912) [inline picard/metadata.py]: It might be possible that total tracks was taken from file metadata and while comparing, the compared release does not have a media tag. (I had an error while testing, which is why I add this)

File: picard/mbjson.py (hunk 12 above)
Comment (samj1912) [inline picard/mbjson.py]: You mean `if obj is not None`?

File: picard/metadata.py (hunk 4 above)
Comment (samj1912) [inline picard/metadata.py]: cases like these were why I added the `if not Track` part

File: picard/mbjson.py (hunk 12 above)
Comment (mineo) [inline picard/mbjson.py]: Yes.

File: picard/metadata.py (hunk 13 above)
Comment (mineo) [inline picard/metadata.py]: Can you reproduce this and the other errors you got during testing? Since the safeguards were not needed for XML, this suggests that the JSON output it not equivalent to the XML one and its documentation needs updating.

File: picard/album.py (hunk 9 above)
Comment (mineo) [inline picard/album.py]: The only ticket related to strings and integers I could find was https://tickets.metabrainz.org/browse/MBS-8329 which appears to be fixed for the most part, are there more?

File: picard/metadata.py (hunk 13 above)
Comment (samj1912) [inline picard/metadata.py]: Unable to reproduce it sorry :(

Comment (samj1912) [inline picard/metadata.py]: I changed the way this is handled in the code though

File: picard/album.py (hunk 9 above)
Comment (samj1912) [inline picard/album.py]: okay, I can still see score is a string. I will remove the type casts then.

File: picard/ui/searchdialog.py [hunk 16]
```diff
@@ -645,19 +645,20 @@ def _cover_downloaded(self, row, data, http, error):
             try:
                 pixmap.loadFromData(data)
                 cover_cell.update(pixmap)
-            except:
+            except Exception as e:
                 cover_cell.not_found()
+                log.error(e)
```
Comment (mineo) [inline picard/ui/searchdialog.py]: `log` is not imported in this file.

Comment (samj1912) [inline picard/ui/searchdialog.py]: fixed
